lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -2501,6 +2508,54 @@ void ModuleEmitter::emitHWGeneratedModule(HWModuleGeneratedOp module) {
   os << "// external generated module " << verilogName.getValue() << "\n\n";
 }
 
+void ModuleEmitter::emitBind(BindOp bind) { os << "// bind\n\n"; }
+
+static bool isSimpleReadOrPort(Value v) {
+  if (v.isa<BlockArgument>())
+    return true;
+  auto read = dyn_cast_or_null<ReadInOutOp>(v.getDefiningOp());
+  if (!read)
+    return false;
+  return dyn_cast_or_null<WireOp>(read.input().getDefiningOp()) ||
+         dyn_cast_or_null<RegOp>(read.input().getDefiningOp());
+}
+
+// Given an invisible instance, make sure all inputs are driven from
+// wires or ports.
+static void lowerBoundInstance(InstanceOp op) {
+  Block *block = op->getParentOfType<HWModuleOp>().getBodyBlock();
+  auto builder = ImplicitLocOpBuilder::atBlockBegin(op.getLoc(), block);
+
+  SmallString<32> nameTmp;
+  nameTmp = (op.instanceName() + "_").str();
+  auto namePrefixSize = nameTmp.size();
+
+  size_t nextOpNo = 0;
+  for (auto &port : getModulePortInfo(op.getReferencedModule())) {
+    if (port.isOutput())
+      continue;
+
+    auto src = op.getOperand(nextOpNo);
+    ++nextOpNo;
+
+    if (isSimpleReadOrPort(src))
+      continue;
+
+    nameTmp.resize(namePrefixSize);
+    if (port.name)
+      nameTmp += port.name.getValue().str();
+    else
+      nameTmp += std::to_string(nextOpNo - 1);
+
+    auto newWire = builder.create<WireOp>(src.getType(), nameTmp);
+    auto newWireRead = builder.create<ReadInOutOp>(newWire);
+    auto connect = builder.create<ConnectOp>(newWire, src);
+    newWireRead->moveBefore(op);
+    connect->moveBefore(op);
+    op.setOperand(nextOpNo - 1, newWireRead);
+  }
```

Comment:
It seems fishy to me that HWModule puts operands and results all in to a single ModulePortInfo array.   I think it would make more sense to split this in to `getOperandInfo()` and `getResultInfo()`. This is probably outside the scope of this change, but i'm curious what you think.

Reply:
I'm not entirely sure about how `HWModule` works, but this could be to interleave operands and results such that the outputs don't all bunch up at the end of the port list?

Reply:
It's not a bad idea.  It is pretty common to have to track which input/output you are at by counting from the start of the array.